### ISSUE TITLE
FileBasedSource does not retry previous WorkUnits unless there is new data

### ISSUE THREAD

**Issue by kadaan** [linkedin/gobblin#556]
The `FileBasedSource.getWorkUnits` method has the capability to include failed WorkUnits from a previous run in the current run.  Unfortunately, these do not get included unless there are new files in the source.  The reason is the following code:

```
List<WorkUnitState> previousWorkunits = Lists.newArrayList(state.getPreviousWorkUnitStates());
List<String> prevFsSnapshot = Lists.newArrayList();

// Get list of files seen in the previous run
if (!previousWorkunits.isEmpty() && previousWorkunits.get(0).getWorkunit().contains(ConfigurationKeys.SOURCE_FILEBASED_FS_SNAPSHOT)) {
    prevFsSnapshot = previousWorkunits.get(0).getWorkunit().getPropAsList(ConfigurationKeys.SOURCE_FILEBASED_FS_SNAPSHOT);
}

// Get list of files that need to be pulled
List<String> currentFsSnapshot = this.getcurrentFsSnapshot(state);
List<String> filesToPull = Lists.newArrayList(currentFsSnapshot);
filesToPull.removeAll(prevFsSnapshot);

List<WorkUnit> workUnits = Lists.newArrayList();
if (!filesToPull.isEmpty()) {
  return workUnits;
}
```

I believe a better solution would be to include the previously failed WorkUnits regardless of whether there are new files available.

**Comment by kadaan** [comment]
@liyinan926 What is your take on this?

**Comment by liyinan926** [comment]
@kadaan I agree with you that failed `WorkUnit`s should be included for retry regardless of whether there are new files to pull.

**Comment by kadaan** [comment]
@liyinan926 I will submit a PR today.

**Comment by liyinan926** [comment]
@kadaan thanks!

**Comment by kadaan** [comment]
Fixed in: #559
